src/Interpreters/PeriodicLog.cpp

```diff
@@ -0,0 +1,62 @@
+#include <Interpreters/PeriodicLog.h>
+#include <Interpreters/ErrorLog.h>
+#include <Interpreters/MetricLog.h>
+
+namespace DB
+{
+
+template <typename LogElement>
+void PeriodicLog<LogElement>::startCollect(size_t collect_interval_milliseconds_)
+{
+    collect_interval_milliseconds = collect_interval_milliseconds_;
+    is_shutdown_metric_thread = false;
+    flush_thread = std::make_unique<ThreadFromGlobalPool>([this] { threadFunction(); });
+}
+
+template <typename LogElement>
+void PeriodicLog<LogElement>::stopCollect()
+{
+    bool old_val = false;
+    if (!is_shutdown_metric_thread.compare_exchange_strong(old_val, true))
+        return;
+    if (flush_thread)
+        flush_thread->join();
+}
+
+template <typename LogElement>
+void PeriodicLog<LogElement>::shutdown()
+{
+    stopCollect();
+    this->stopFlushThread();
```

Comment:
Even though the `this` looks a little weird, this is to prevent a compilation error:

>error: use of undeclared identifier 'stopFlushThread'

An alternative would be to use

```c
SystemLog<LogElement>::stopFlushThread()
```
but it's much more verbose